File: cogs/general.py
import datetime
import os
import time

import discord
from discord import app_commands
from discord.ext import commands

from core.config import BOT_INVITE_PERMISSIONS
from core.issues import GitHubIssueModal


class General(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot
        self.start_time = time.time()

    @app_commands.command(name="bug", description="Report a bug to the developers.")
    async def bug_report(self, interaction: discord.Interaction):
        """Report a bug to the developers."""
        await interaction.response.send_modal(GitHubIssueModal(issue_type="bug"))

    @app_commands.command(
        name="featurerequest", description="Request a new feature for the bot."
    )
    async def feature_request(self, interaction: discord.Interaction):
        """Request a new feature for the bot."""
        await interaction.response.send_modal(GitHubIssueModal(issue_type="feature"))

    @commands.command()
    async def invite(self, ctx: commands.Context) -> None:
        """Get the bot invite URL with the configured permissions (for adding the bot to a server)."""
        app_id = self.bot.application_id or os.getenv("DISCORD_APPLICATION_ID")
        if not app_id:
            await ctx.send(
                "❌ Application ID not available. Set DISCORD_APPLICATION_ID in your environment."
            )
            return
        app_id = int(app_id) if isinstance(app_id, str) else app_id
        permissions = discord.Permissions(BOT_INVITE_PERMISSIONS)
        url = discord.utils.oauth_url(app_id, scopes=["bot"], permissions=permissions)
        await ctx.send(f"**Add this bot to a server:**\n{url}")

    @commands.command()
    async def status(self, ctx: commands.Context) -> None:
        """Check the bot's status and uptime."""
        uptime_seconds = int(time.time() - self.start_time)
        uptime_str = str(datetime.timedelta(seconds=uptime_seconds))

        embed = discord.Embed(title="Bot Status", color=discord.Color.green())
        embed.add_field(name="Uptime", value=uptime_str, inline=True)
        embed.add_field(
            name="Ping", value=f"{round(self.bot.latency * 1000)}ms", inline=True
        )

        try:
            load1, load5, load15 = os.getloadavg()
            embed.add_field(
                name="System Load",
                value=f"{load1:.2f}, {load5:.2f}, {load15:.2f}",
                inline=False,
            )
        except OSError:
            pass

        embed.set_footer(text=f"Server ID: {ctx.guild.id if ctx.guild else 'DM'}")
        await ctx.send(embed=embed)

    @commands.Cog.listener()
    async def on_voice_state_update(
        self,
        member: discord.Member,
        before: discord.VoiceState,
        after: discord.VoiceState,
    ) -> None:
        """Automatically disconnects the bot if it's the only one left in the voice channel."""
        voice_client = member.guild.voice_client
        if (
            voice_client
            and voice_client.channel
            and len(voice_client.channel.members) == 1
        ):
            await voice_client.disconnect(force=False)


async def setup(bot: commands.Bot):
    await bot.add_cog(General(bot))

File: packages/bot/src/commands/general.ts
import * as os from 'os';
import * as config from '../core/config.js';
import type { AffixDisplay } from '@mythicplus/shared';

export interface GeneralContext {
  guild: { id: string } | null;
  send(content: string, options?: Record<string, unknown>): Promise<unknown>;
}

export interface VoiceClient {
  channel: { members: unknown[] };
  disconnect(options?: { force?: boolean }): Promise<void>;
}

export interface EmbedData {
  title: string;
  color: number;
  fields: { name: string; value: string; inline?: boolean }[];
  footer?: { text: string };
}

export class GeneralHandler {
  private startTime: number;
  latency: number;
  applicationId?: string | null;

  constructor(latency = 0, applicationId?: string | null) {
    this.startTime = Date.now() / 1000;
    this.latency = latency;
    this.applicationId = applicationId;
  }

  /** Override start time for testing. */
  _setStartTime(t: number): void {
    this.startTime = t;
  }

  private getCommitValue(): string {
    if (config.GIT_SHA && config.GIT_SHA.length >= 7) {
      const shortSha = config.GIT_SHA.slice(0, 7);
      const commitUrl = `https://github.com/${config.GITHUB_REPO_OWNER}/${config.GITHUB_REPO_NAME}/commit/${config.GIT_SHA}`;
      return `[\`${shortSha}\`](${commitUrl})`;
    }
    return 'unknown';
  }

  async version(ctx: GeneralContext): Promise<void> {
    const embed: EmbedData = {
      title: 'Bot Version',
      color: 0x3498db,
      fields: [{ name: 'Commit', value: this.getCommitValue(), inline: false }],
    };
    await ctx.send('', { embed });
  }

  async invite(ctx: GeneralContext): Promise<void> {
    const appId = this.applicationId ?? config.DISCORD_APPLICATION_ID;
    if (!appId) {
      await ctx.send(
        '❌ Application ID not available. Set config.DISCORD_APPLICATION_ID in your environment.',
      );
      return;
    }
    const permissions = config.BOT_INVITE_PERMISSIONS;
    const url = `https://discord.com/api/oauth2/authorize?client_id=${appId}&scope=bot&permissions=${permissions}`;
    await ctx.send(`**Add this bot to a server:**\n${url}`);
  }

  async status(ctx: GeneralContext): Promise<void> {
    const now = Date.now() / 1000;
    const uptimeSeconds = Math.floor(now - this.startTime);

    const hours = Math.floor(uptimeSeconds / 3600);
    const minutes = Math.floor((uptimeSeconds % 3600) / 60);
    const seconds = uptimeSeconds % 60;
    const uptimeStr = `${hours}:${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;

    const pingMs = Math.round(this.latency * 1000);

    const embed: EmbedData = {
      title: 'Bot Status',
      color: 0x2ecc71,
      fields: [
        { name: 'Uptime', value: uptimeStr, inline: true },
        { name: 'Ping', value: `${pingMs}ms`, inline: true },
        { name: 'Commit', value: this.getCommitValue(), inline: true },
      ],
    };

    try {
      const loadavg = os.loadavg();
      embed.fields.push({
        name: 'System Load',
        value: `${loadavg[0].toFixed(2)}, ${loadavg[1].toFixed(2)}, ${loadavg[2].toFixed(2)}`,
        inline: false,
      });
    } catch {
      // loadavg not available on some platforms
    }

    const serverId = ctx.guild?.id ?? 'DM';
    embed.footer = { text: `Server ID: ${serverId}` };

    await ctx.send('', { embed });
  }

  async affixes(ctx: GeneralContext, affixes: AffixDisplay[]): Promise<void> {
    const colorDot: Record<string, string> = {
      '#22c55e': '🟢',
      '#a855f7': '🟣',
      '#ef4444': '🔴',
      '#f59e0b': '🟡',
    };

    const lines = affixes.map((a) => {
      const dot = colorDot[a.color?.toLowerCase()] ?? '⚪';
      const nick = a.nickname ? ` — *${a.nickname}*` : '';
      return `${dot} [**${a.name}**](${a.wowheadUrl}) (${a.keystoneLevel})${nick}`;
    });

    const embed: EmbedData = {
      title: "This Week's Affixes",
      color: 0x9146ff,
      fields: [{ name: '\u200b', value: lines.join('\n') || 'No affixes available.', inline: false }],
    };
    await ctx.send('', { embed, ephemeral: true });
  }

  async onVoiceStateUpdate(voiceClient: VoiceClient | null): Promise<void> {
    if (voiceClient && voiceClient.channel && voiceClient.channel.members.length === 1) {
      await voiceClient.disconnect({ force: false });
    }
  }
}
